MetaMorpheus/EngineLayer/DIA/DIAEngine.cs

```diff
@@ -0,0 +1,147 @@
+        public List<Ms2ScanWithSpecificMass> PseudoMs2Scans { get; set; } 
+        protected override MetaMorpheusEngineResults RunSpecific()
+        {
+            //read in scans
+            var ms1Scans = DataFile.GetMS1Scans().ToArray();
+            var ms2Scans = DataFile.GetAllScansList().Where(s => s.MsnOrder == 2).ToArray();
+            var DIAScanWindowMap = ConstructMs2Groups(ms2Scans);
+
+            //Get all MS1 and MS2 XICs
+            var allMs1Xics = new ConcurrentDictionary<(double min, double max), List<ExtractedIonChromatogram>>();
+            var allMs2Xics = new ConcurrentDictionary<(double min, double max), List<ExtractedIonChromatogram>>();
+            Parallel.ForEach(DIAScanWindowMap, new ParallelOptions { MaxDegreeOfParallelism = CommonParameters.MaxThreadsToUsePerFile }, ms2Group =>
+                {
+                    allMs1Xics[ms2Group.Key] = DIAparams.Ms1XicConstructor.GetAllXics(ms1Scans, new MzRange(ms2Group.Key.min, ms2Group.Key.max));
+                    allMs2Xics[ms2Group.Key] = DIAparams.Ms2XicConstructor.GetAllXics(ms2Group.Value.ToArray());
+                });
+
+            //Precursor-fragment Grouping
+            var allPfGroups = new List<PrecursorFragmentsGroup>();
+            foreach (var ms2Group in DIAScanWindowMap.Keys)
+            {
+                var pfGroups = DIAparams.PfGroupingEngine.PrecursorFragmentGrouping(allMs1Xics[ms2Group], allMs2Xics[ms2Group]);
+                allPfGroups.AddRange(pfGroups);
+            }
+
+            //Convert pfGroups to pseudo MS2 scans
+            PseudoMs2Scans = new List<Ms2ScanWithSpecificMass>();
+            int pfGroupIndex = 1;
+            foreach (var pfGroup in allPfGroups)
+            {
+                pfGroup.PFgroupIndex = pfGroupIndex;
+                var pseudoScan = PrecursorFragmentsGroup.GetPseudoMs2ScanFromPfGroup(pfGroup, DIAparams.PseudoMs2ConstructionType, CommonParameters, DataFile.FilePath);
+                PseudoMs2Scans.Add(pseudoScan);
+                pfGroupIndex++;
+            }
+            return new MetaMorpheusEngineResults(this);
+        }
+
+        public DIAEngine(DIAparameters DIAparameters, MsDataFile dataFile, CommonParameters commonParameters, List<(string FileName, CommonParameters Parameters)> fileSpecificParameters, List<string> nestedIds) :base(commonParameters, fileSpecificParameters, nestedIds)
+        {
+            DIAparams = DIAparameters;
+            DataFile = dataFile;
+        }
+
+        /// <summary>
+        /// Divide all MS2 scans into groups based on their isolation windows. 
+        /// </summary>
+        /// <returns> A dictionary where the key is (minumum of isolation window, maximum of isolation window), 
+        /// and the values are MS2 scans that belongs to this isolation window. </returns>
+        public static Dictionary<(double min, double max), List<MsDataScan>> ConstructMs2Groups(MsDataScan[] ms2Scans, int binSize = 100, double tolerance = 5)
```

Comment:
Check out the DoubleRange class in MzLib as well as the MzRange class. They have some useful methods that could be used here and elsewhere instead of the tuple.